feat: dashboard <> insight access control checks

zlwaterfield opened the pull request:
## Changes


The core idea
- insights and dashboards have their own access controls
- a user can have access to a dashboard but might not have access to all insights on it, so this is making sure we have the proper checks for that
- if a user has view access to the dashboard, they will be able to view it and not edit anything about the dashboard, the checks for the insights on if are done separately

Improving the checks for access controls on dashboards and insights.
1. Check insight access controls server side and don't load data/details if the user doesn't have view access.
2. On the dashboard if the user doesn't have view access return a custom card meta and insight error state to block access
3. Stop the insight logic from refreshing the insight data if the user doesn't have view access
4. Improve the edit access checks in the actions dropdown

Here's is how an insight will show up:
<img width="650" height="502" alt="Screenshot 2025-09-11 at 4 58 13 PM" src="https://github.com/user-attachments/assets/e833b319-c802-44a2-99b2-2d0300c84cff" />



## How did you test this code?

Manually

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

zlwaterfield committed:
Simplify permission warning in access control panel

zlwaterfield committed:
Refactor access control handling in LemonButton and LemonSwitch

Replaces manual access level checks and disabled reason construction with the new getAccessControlDisabledReason utility in both LemonButton and LemonSwitch components. This centralizes and simplifies access control logic and messaging.

zlwaterfield committed:
Add option to omit access details in disabled reason

Updated getAccessControlDisabledReason to accept an optional includeAccessDetails parameter. This allows callers to control whether detailed access level information is included in the returned message.

zlwaterfield committed:
Add access control check to InsightCard view

Introduces a check for user access level before rendering insight content in InsightCard. If the user lacks permission, an error state is displayed, improving security and user feedback.

zlwaterfield committed:
Add access control check to getInsightWithRetry

Before making API calls, getInsightWithRetry now verifies if the user has viewer access to the insight. If access is insufficient, the function returns the insight as-is, preventing unnecessary API requests.

zlwaterfield committed:
Restrict insight data for insufficient dashboard access

Adds a check in InsightSerializer to return only minimal insight data when a user views an insight in a dashboard context without sufficient access level. This helps enforce RBAC by limiting data exposure based on user permissions.

zlwaterfield committed:
Add access control checks to InsightMeta component

Replaces privilege level checks with access control utility functions to determine view and edit permissions for insights. If the user lacks view access, a minimal interface with an 'Access denied' message is shown. Edit-related UI elements are now conditionally rendered based on the new access checks.

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

zlwaterfield committed:
Apply suggestion from @greptile-apps[bot]

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

zlwaterfield committed:
Refine dashboard and insight permission checks

Updated CardMeta to make moreButtons optional. In InsightMeta, separated dashboard and insight permission checks, ensuring dashboard actions require dashboard edit access and insight deletion requires insight edit access. Also fixed import for AccessControlLevel in dashboardUtils.

zlwaterfield committed:
Merge branch 'zach/dashboard-insight-ac-improvements' of github.com:PostHog/posthog into zach/dashboard-insight-ac-improvements

zlwaterfield committed:
Add access control stories and improve access checks

Introduces Storybook stories for InsightCard and Dashboard components to demonstrate different access control scenarios. Updates access level checks throughout InsightCard, InsightMeta, and dashboardUtils to handle insights without explicit user_access_level, defaulting to permissive behavior for legacy insights.

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

zlwaterfield committed:
Merge branch 'zach/dashboard-insight-ac-improvements' of github.com:PostHog/posthog into zach/dashboard-insight-ac-improvements

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

thmsobrmlr reviewed (approved):
Code looks good and works great!

Minor thing I found is that loading an insight with "no access" results in an infinite loading state (since we fetch it from the list endpoint and not the single insight endpoint). Don't think this needs to be fixed now.

Mocks for snapshots need to be adapted though.

zlwaterfield committed:
Merge remote-tracking branch 'origin/master' into zach/dashboard-insight-ac-improvements

zlwaterfield committed:
Merge branch 'master' into zach/dashboard-insight-ac-improvements

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

zlwaterfield committed:
Merge branch 'master' into zach/dashboard-insight-ac-improvements

zlwaterfield committed:
clear snapshots so they are regenerated

zlwaterfield committed:
Merge branch 'zach/dashboard-insight-ac-improvements' of github.com:PostHog/posthog into zach/dashboard-insight-ac-improvements

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

zlwaterfield committed:
ignore lint

zlwaterfield committed:
Add user access control to dashboard serializer test

Introduces UserAccessControl to the dashboard serializer context in the test to ensure proper user permissions are handled during serialization.
